core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -350,20 +350,16 @@ private[spark] class TaskSchedulerImpl(
                 removeExecutor(execId, reason.get)
                 failedExecutor = Some(execId)
               }
+            }
+            if (TaskState.isFinished(state)) {
+              cleanupTaskState(tid)
```

Comment:
Fair enough, although I ended up preserving the ordering just to reduce the set of changes that we need to reason about.